Custom workflows plan

veerajsk opened the pull request:
PopulateBlackboard:

- Adds a PopulateBlackboard plan implementation in the alica_standard_library which populates all protected blackboard keys from the input key `data` which is assumed to be json formatted. The type is taken from the blackboard blueprint & the json is parsed to set the value. 

- i.e. it effectively does `blackboard.set("protected_key", YAML::Load(blackboard.get<std::string>("data")).as<type_from_blueprint>()` for every protected key.

- If a key is not present in the input `data` it is skipped

- adds a unit test for testing PopulateBlackboard

CustomWorkflowTutorial:

- The PopulateBlackboard implementation is used to demonstrate how to create custom workflows by adding a tutorial in turtlesim.
- The tutorial makes use of PopulateBlackboard to parse & populate input keys for sub plans from a task msg published on a topic

Engine changes:

- The Parser in blackboard.h now simply forwards string values instead of calling Load on them. This is required to pass `json` values as strings instead of parsing them again which was resulting in incorrect behaviour.
- Need to read the access type of a blackboard value into the blueprint
- Provide access to the blackboard blueprint via BasicPlan & BasicBehaviour

Note: Change seems big mainly because of plan changes i.e. due to added/modified pml & beh files etc.

veerajsk committed:
Implementation

veerajsk committed:
Merge branch 'devel' into vsk_populate_blackboard_beh

veerajsk committed:
Move files

veerajsk committed:
Include fix

veerajsk committed:
Merge branch 'devel' into vsk_populate_blackboard_beh

veerajsk committed:
Fix file name typo

veerajsk committed:
Set success, failure

veerajsk committed:
Add tutorials

veerajsk committed:
Fix file & namespace name

veerajsk committed:
Implement plans

veerajsk committed:
Merge branch 'devel' into vsk_populate_blackboard_beh

veerajsk committed:
Fixes

veerajsk committed:
Merge branch 'devel' into vsk_populate_blackboard_beh

veerajsk committed:
Merge branch 'devel' into vsk_custom_workflows_plan

veerajsk committed:
Fix plans

veerajsk committed:
Implement populate blackboard

veerajsk committed:
working version

veerajsk committed:
Merge branch 'devel' into vsk_custom_workflows_plan

veerajsk committed:
Unit test

veerajsk committed:
Fix parsing

veerajsk committed:
Remove unused condition

veerajsk committed:
Delete unused file

veerajsk committed:
Remove unused function

veerajsk committed:
Remove unused files

veerajsk committed:
Fix supplementary tests

Dale-Koenig commented on alica_standard_library/include/PopulateBlackboard.h:
```diff
@@ -0,0 +1,56 @@
+    template <std::size_t... Is>
+    void setHelper(const std::string& key, std::size_t typeIndex, std::index_sequence<Is...>)
+    {
+        struct Dummy
```
Can you comment to explain what this is doing?

veerajsk replied:
this is just syntactic sugar to call `setIfIndex` for every index in the index_sequence.
The comma operator makes sure `setIfIndex` is called, however the value of the expression `(setIfIndex, 0)` itself is `0`, so basically `dummyFunc` will be called with a bunch of 0s

Dale-Koenig commented:
>i.e. it effectively does blackboard.set("protected_key", YAML::Load(blackboard.get<std::string>("data")).as<type_from_blueprint>() for every protected key.

We only want to do this for every key in the json I think.  It is possible there are protected keys that are not meant to be set from the json

bjoernschroeder commented:
> We only want to do this for every key in the json I think. It is possible there are protected keys that are not meant to be set from the json

But all protected keys which are not found in the json file are skipped. I dont think there should be any case where we have the key as protected in the blueprint and the same key in the json file and we dont want to set it, right?

veerajsk commented:
> > i.e. it effectively does blackboard.set("protected_key", YAML::Load(blackboard.getstd::string("data")).as<type_from_blueprint>() for every protected key.
> 
> We only want to do this for every key in the json I think. It is possible there are protected keys that are not meant to be set from the json

yeah, that's the reason that if the key is not found in the json, we skip over to the next key without raising any kind of error. Also I think we don't require to map every key in the json to some key on the blackboard either, so there is no error raised for that either

veerajsk commented:
> > We only want to do this for every key in the json I think. It is possible there are protected keys that are not meant to be set from the json
> 
> But all protected keys which are not found in the json file are skipped. I dont think there should be any case where we have the key as protected in the blueprint and the same key in the json file and we dont want to set it, right?

so, a key will be set if it is in the json & in the blackboard & it is protected.

veerajsk committed:
Replace transport workflow with teleport workflow

Transport workflow does not make much sense for turtles therefore
replace the tutorial with a teleport workflow instead

Dale-Koenig commented:
Looks like some of the blackboard mappings in the tutorial are broken now

veerajsk committed:
Fix mapping

veerajsk commented:
> Looks like some of the blackboard mappings in the tutorial are broken now

strange, I tested the changes as well. Not sure how this happened. I fixed it though & tested again
